### ISSUE TITLE
Add support for connecting the topology of polygon-soup (Merge algorithm)

### ISSUE THREAD

**Issue by sebjf** [3drepo/pmp-library#3]
Some meshes, especially those with per-vertex normals created by naive importers, contain many borders, or in the worst case polygon-soups consisting of entirely separate triangles.

These meshes may look fully connected, but topologically consist of separate faces.

![image](https://user-images.githubusercontent.com/5058369/227529021-87f5456d-d21c-4630-86ce-c52d14a996a8.png)

This is a problem for algorithms like Decimation algorithms, for which connectivity is an important source of information.

An algorithm should be added to automatically re-topologise meshes with coincident vertices (with tuneable parameters), to prepare meshes for Decimation.

The first step of re-topologising such meshes is to combine coincident vertices based on some criteria (such as the normals being within a given angle). Currently, `SurfaceMesh` supports this by changing the `to` vertex of a `HalfEdge`.

![image](https://user-images.githubusercontent.com/5058369/227529382-80c06a69-826e-4f12-a9ea-5a8b21128a90.png)

Once this is done however, the topology is still not connected because there still boundary edges, it just they now share the same vertex references. To connect the topology, a second stage should search for these edges and combine them into one, connecting incident faces.

This is done in by creating a hash of all half-edges in the mesh, and filtering those that can be combined (i.e. where edges mirror each other across a boundary). Then, one edge can be merged into another by adopting its topology. The surrounding boundary edges are re-directed. Finally the edge merged from is isolated and deleted.

![image](https://user-images.githubusercontent.com/5058369/227530964-36954929-666f-4187-aea6-411ee01565e2.png)

(In the image above, the edge containing h2 is merged into the edge containing h1. h1 adopts h2's connectivity, and the edge containing h2 is deleted).
